### ISSUE TITLE
Middleware returns int(1)

### ISSUE THREAD

**Issue by realart** [slimphp/Slim#1203]
Disclaimer: Most likely not an issue, just lack of insight on my part.

I am using v3 development version and my project relies on things like:
- Subrequests (killer feature, makes it a HMVC framework)
- Application-wide middleware for things like wrapping page routes inside layout views etc.

All was fine and dandy until I updated to latest commit of the develop branch, and middleware calls started to act strange:
- Return integer 1, instead of Response object
- Leak output

For example, a call like this "$newResponse = $next($req, $res)" inside application middleware would:
- Return integer 1 as the "$newResponse" variable
- Dump the output (the contents of what should have been response body) of the "$next" callable to the screen

**Comment by akrabat** [comment]
Are you able to create a small example project that shows the problem?

**Comment by realart** [comment]
If I invoke /test, it prints "some nice output" and crashes with exception:

```
UnexpectedValueException: Middleware must return instance of \Psr\Http\Message\ResponseInterface'

#0 C:\projects\flsite\vendor\slim\slim\Slim\MiddlewareAware.php(106): Slim\App->Slim\{closure}(Object(Slim\Http\Request), Object(Slim\Http\Response))
#1 C:\projects\flsite\vendor\slim\slim\Slim\App.php(346): Slim\App->callMiddlewareStack(Object(Slim\Http\Request), Object(Slim\Http\Response))
#2 C:\projects\flsite\index.php(16): Slim\App->run()
#3 {main}
```

Index.php

```
require 'vendor/autoload.php';

$app = new \Slim\App();

$app->get('/test', function($req, $res, $args) {
    echo 'some nice output';
});

$app->add(function($req, $res, $next) use($app) {
    $newResponse = $next($req, $res);
    return $newResponse;
});

$app->run();
```

**Comment by akrabat** [comment]
You have to return a Response from the route's action callable:

```
$app->get('/test', function($req, $res, $args) {
   return $response->write('some nice output');
});
```

**Comment by realart** [comment]
But this will force me to manually do output buffering (start my callable with ob_start and end with writing contents of ob_get_clean into response body) in each and every route callable. Previously this was handled automatically by framework.

My applications route definitions are structured like this:

```
$app->get('/', function($req, $res, $args) { return include('pages/home.php'); });
$app->get('/search', function($req, $res, $args) { return include('pages/search.php'); });
$app->get('/fragments/users', function($req, $res, $args) { return include('fragments/users.php'); });
etc.
```

Where each included file handles everything (setting page title, invoking 404 handler, outputting response body with echo or plain text...)

# How I use Slim 3.0

I am not using templating engine. Include files serve both controller and view functionality with plain PHP. This approach actually releaves me of unnecesary overhead of building variable arrays for passing to template object etc. And, believe or not, I am actually somewhat of  a MVC purist :D

I have two types of routes:
- "page routes" which represent a full page URL that visitor may visit
- "fragment routes" which output HTML of some fragment of a page (think "partial views"). These are used heavily with Ajax for various widgets, partial loads etc.

I use one application middleware callable which looks at the request URL and, if it does NOT start with "/fragments/", calls the next middleware (which corresponds to the actual page route callabe), saves its response body in a variable and pastes that variable inside a layout view (think "master view"), which, itself is just a plain PHP template and contains the head and body structure of HTML document, site-wide javascript and css references etc.

For both page and fragment routes, my single application middleware callable also looks up if there is a corresponding javascript file in my sites js folder (with the same name as the route fragment) and includes it.

I can also output response of any fragment route inside any other route using subRequest.

**Comment by akrabat** [comment]
Fixed in #1206.
